packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/meta/DependencyInjection.kt

```diff
@@ -0,0 +1,135 @@
+// be straightforward (to avoid lag) and because state updates should be sequential.
+private val asyncStatusRefreshScope = Dispatchers.IO.limitedParallelism(1)
+
+class AsyncState<E : Any, T : Any>(
+    parent: Disposable,
+    private val sharedFlow: SharedFlow<E>,
+    private val accessor: E.() -> T,
+    private val onChange: suspend T.() -> Unit,
+) : Disposable {
+    private lateinit var state: T
+    private val scope: CoroutineScope = CoroutineScope(asyncStatusRefreshScope)
+
+    init {
+        scope.launchChildBackground {
+            sharedFlow.collectLatest {
+                val newState = it.accessor()
+                if (state != newState) {
+                    state = newState
+                    scope.launchChildBackground {
+                        state.onChange()
+                    }
+                }
+            }
+        }
+
+        Disposer.register(parent, this)
+    }
+
+    operator fun getValue(
+        thisRef: Any?,
+        property: KProperty<*>
+    ): T {
+        return state
+    }
+
+    override fun dispose() {
+        scope.cancel()
+    }
+}
+
+/**
+ * We will need to synchronise the state from a SharedFlow when we are doing work in the background.
+ * However, we want to make it seamless, and we don't want to by coupled to the shared flow syntax.
+ * This delegate should synchronise the variable with the latest event of the SharedFlow. For example:
+ *
+ * ```kt
+ * class EventNumber(val numberInEvent: Int)
+ * val sharedFlow: SharedFlow<EventNumber>
+ * val lastNumber by sharedFlow.latest { numberInEvent }
+ * ```
+ *
+ * It can also be used to compute an expression across properties in an event:
+ * ```kt
+ * class EventNumber(val numberInEvent: Int, val otherNumber: Int)
+ * val sharedFlow: SharedFlow<EventNumber>
+ * val computedNumber by sharedFlow.latest { numberInEvent * otherNumber }
+ * ```
+ *
+ * We can also react to changes if we provide an onChange callback. This callback will run in
+ * another coroutine and is a suspend function, so it can run more heavy functions.
+ *
+ * ```kt
+ * class EventNumber(val numberInEvent: Int, val otherNumber: Int)
+ * val sharedFlow: SharedFlow<EventNumber>
+ * val computedNumber by sharedFlow.latest(
+ *      onNewState = { numberInEvent * otherNumber },
+ *      onChange = {
+ *          val query = queryMongoDb(this) // this = the new state
+ *          setComboBoxModel(query)
+ *      }
+ * )
+ * ```
+ *
+ * It will be closed when the current project is closed. If no projects are available,
+ * it will be closed when the IDE is closed.
+ */
+fun <E : Any, T : Any> SharedFlow<E>.latest(
+    parent: Disposable = ProjectManager.getInstance().openProjects.firstOrNull()
+        ?: ApplicationManager.getApplication(),
+    onNewState: E.() -> T,
+    onChange: suspend T.() -> Unit = {}
+
+): AsyncState<E, T> =
+    AsyncState(
+        parent,
+        this,
+        onNewState,
+        onChange
+    )
```

Comment:
I am gonna try to use this in ToolbarModel, its exactly what we need for those reactive UI states. Thanks for this <3